From afdc6997f23baa9112bc309c8e94d744144eb1b0 Mon Sep 17 00:00:00 2001
From: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Date: Fri, 14 Sep 2018 23:30:57 -0700
Subject: [PATCH 0400/1852] xilinx: v4l: dma: Add DMA input direction support

commit 609386066b8c165a5b0794706cf2e4ed86a632d0 from
https://github.com/Xilinx/linux-xlnx.git

This patch adds DMA input direction support. The logic will understand
the v4l2 device type based on v4l2 capabilities and performs the checks
to support the direction.

This will walk through the subdevices multiple times in case of mem2mem
pipeline as shown below, nevertheless the current implementation maintains
streamed subdevices and there won't be any issues.

DMA1 -> subdev -> DMA2

Signed-off-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 9a659aaf0ab0..4d6c2848225f 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -150,8 +150,19 @@ static int xvip_pipeline_start_stop(struct xvip_composite_device *xdev,
 		if (!pad)
 			break;
 
-		if (!(pad->flags & MEDIA_PAD_FL_SINK))
-			break;
+		/*
+		 * This will walk through the subdevices multiple times in case
+		 * of mem2mem pipeline. But there won't be any issues as
+		 * driver is maintaining list of streamed subdevices
+		 */
+		if (xdev->v4l2_caps & V4L2_CAP_VIDEO_CAPTURE_MPLANE ||
+		    xdev->v4l2_caps & V4L2_CAP_VIDEO_CAPTURE) {
+			if (!(pad->flags & MEDIA_PAD_FL_SINK))
+				break;
+		} else {
+			if (!(pad->flags & MEDIA_PAD_FL_SOURCE))
+				break;
+		}
 
 		pad = media_entity_remote_pad(pad);
 		if (!pad || !is_media_entity_v4l2_subdev(pad->entity))
-- 
2.31.1

